File: src/mission/tasks/route.cpp
#include "../../nodes.h"
#include "../../comms/events.h"
#include "../../fcs/fcs_mgr.h"
#include "../mission_mgr.h"
#include "route.h"

route_task_t::route_task_t(PropertyNode config_node) {
    name = "route";
}

void route_task_t::activate() {
    active = true;

    // defaults
    route_node.setString("follow_mode", "leader");
    route_node.setString("start_mode", "first_wpt");
    route_node.setString("completion_mode", "loop");

    // set modes
    fcs_mgr->set_mode("basic+tecs");
    mission_node.setString("mode", "route");
    event_mgr->add_event("mission", "route");
}

void route_task_t::update(float dt) {
    // route following is one of the two fundamental modes so there is nothing
    // to update here once the task is setup and activated.
}

bool route_task_t::is_complete() {
    return false;
}

void route_task_t::close() {
    active = false;
}

File: src/mission/tasks/tasks-python-archive/route.py
import math

from PropertyTree import PropertyNode

import comms.events
import control.route
from mission.task.task import Task
from mission.task import fcsmode
import mission.task.state

class Route(Task):
    def __init__(self, config_node):
        Task.__init__(self)
        self.route_node = PropertyNode("/task/route")
        self.ap_node = PropertyNode("/autopilot")
        self.nav_node = PropertyNode("/navigation")
        self.targets_node = PropertyNode("/autopilot/targets")

        self.alt_agl_ft = 0.0
        self.speed_kt = 30.0

        self.name = config_node.getString("name")
        self.coord_path = config_node.getString("coord_path")
        self.alt_agl_ft = config_node.getDouble("altitude_agl_ft")
        self.speed_kt = config_node.getDouble("speed_kt")

        # load a route if included in config tree
        if control.route.build(config_node):
            control.route.swap()
        else:
            print("Detected an internal inconsistency in the route")
            print(" configuration.  See earlier errors for details.")
            quit()

    def activate(self):
        self.active = True

        # save existing state
        mission.task.state.save(modes=True, targets=True)

        # set modes
        fcsmode.set("basic+tecs")
        self.nav_node.setString("mode", "route")

        if self.alt_agl_ft > 0.1:
            self.targets_node.setDouble("altitude_agl_ft", self.alt_agl_ft)

        self.route_node.setString("follow_mode", "leader");
        self.route_node.setString("start_mode", "first_wpt");
        self.route_node.setString("completion_mode", "loop");

        comms.events.log("mission", "route")

    # build route from a property tree node
    def build(self, config_node):
        self.standby_route = []       # clear standby route
        num = config_node.getLen("wpt")
        for i in range(num):
            child = config_node.getChild("wpt/%d" % i)
            wp = waypoint.Waypoint()
            wp.build(child)
            self.standby_route.append(wp)
        print("loaded %d waypoints" % len(self.standby_route))
        return True

    def update(self, dt):
        if not self.active:
            return False

    def is_complete(self):
        return False

    def close(self):
        # restore the previous state
        mission.task.state.restore()

        self.active = False
        return True
